src/tests/ftest/control/ms_resilience.py

```diff
@@ -0,0 +1,244 @@
+            # Verify that the pool persisted.
+            pool_data = self.server_managers[-1].dmg.pool_list()
+            if pool_data and self.pool.uuid in pool_data:
+                self.log.info("Found pool in system.")
+            else:
+                self.fail("No pool found in system.")
+
+    def verify_leader(self, access_list):
+        """Verify the leader of the MS is in the access_list.
+
+        Args:
+            access_list (list): list of hostname representing the access points
+                for the MS.
+
+        Returns:
+            str: hostname of the MS leader. If the leader is not within the
+                access_list, the test will fail.
+
+        """
+        l_addr = self.server_managers[-1].dmg.system_leader_query()["leader"]
+        l_hostname, _, _ = socket.gethostbyaddr(l_addr.split(":")[0])
+        l_hostname = l_hostname.split(".")[0]
+
+        # Check that the new leader is in the access list
+        if l_hostname not in access_list:
+            self.fail("Selected leader <{}> is not within the access_list"
+                      " provided to servers".format(l_hostname))
+        return l_hostname
+
+    def launch_servers(self, resilience_num):
+        """Setup and start the daos_servers.
+
+        Args:
+            resilience_num (int): minimum amount of MS replicas to achieve
+                resiliency.
+
+        Returns:
+            list: list of access point hosts where MS has been started.
+
+        """
+        access_list = random.sample(self.hostlist_servers, resilience_num)
+        server_groups = {
+            self.server_group:
+                {
+                    "servers": self.hostlist_servers,
+                    "access_list": access_list
+                },
+        }
+        self.start_servers(server_groups)
+        return access_list
+
+    def verify_resiliency(self, N):
+        """Verify 2N+1 resiliency up to N = 2"""
+        access_list = self.launch_servers((2 * N) + 1)
+        leader = self.verify_leader(access_list)
+
+        # Shutdown the servers that is the leader in the MS
+        stop_processes([leader], "daos_server")
+
+        # Remove the server that was stopped from the access_list
+        access_list = [x for x in access_list if x != leader]
+        self.server_managers[-1].dmg.hostlist = access_list[0]
+
+        # Verify that the MS is still accessible
+        # Let's try to make an update by creating a pool on the group
+        self.verify_leader(access_list)
+        self.update_and_verify(ignore_status=False)
+
+    def verify_lost_resiliency(self, N):
+        """Test that even with 2N+1 resiliency lost, reads still work."""
+        access_list = self.launch_servers((2 * N) + 1)
+        leader = self.verify_leader(access_list)
+
+        # Shutdown N + 1 access_list hosts, including the MS leader
+        kill_list = random.sample(access_list, N)
+        kill_list.append(leader)
```

Comment:
Is it possible that we add duplicate value to kill_list here? If so, does it cause a problem?

Reply:
Hmm. Let me verify that, I'll get back to you.

Reply:
@shimizukko , we should be fine here. random.sample() is sampling without replacement, so we will not get duplicate hosts.